### PR TITLE
String obfuscation in the repo

### PR DESCRIPTION
**Pull request by Jack7D1**
## About The Pull Request
Adds a subsystem and tool pair that can obfuscate strings in the source code that are only made plainly visible at init. Supports default fallback values that will appear in the source code and on servers that the string was not expressly obfusticated for.
Note:
This subsystem is not secure, anyone with the world.url that the string is targeted for can decode it themselves.
The words 'obfuscation' are intentionally used in place of 'encryption', as this only makes it *harder* to read these strings in the source.

## Why It's Good For The Game
This provides a powerful tool for future code to build off of. Maintaining a thin veil of secrecy in the source code.

## Changelog
:cl:
add: Text obfuscation subsystem for hiding strings in the code to be decoded at runtime.
fix: Meatball names are fixed
/:cl:

### CODE REVIEW

**Comment by Superyodeler** [comment]
Do not spook wonerfil raptil

**Comment by Maldaris** [comment]
Would like to go on record as having said: this PR is unnecessary for its stated/implied use on our server. We are an 18+ server and community, which does abide by the freedom of speech.

That being said I do think this has some fun potential for encrypting strings for exploration content. This could be used to mildly obfuscate puzzles and content that would otherwise be readable/complete-able in singleplayer. Very interested to see where this goes.

File: code/game/world.dm [hunk 1]
```diff
@@ -376,3 +376,21 @@ GLOBAL_VAR(restart_counter)
 
 
 /world/proc/refresh_atmos_grid()
+
+//Defining global vars used for handling bad words. Init as a global to save huge processing time.
+/var/global/list/badwordlist[BAD_WORD_COUNT]
```
Comment (Shadowtail117) [inline code/game/world.dm:381]: The game doesn't like when you declare global stuff like this. Use `GLOBAL_LIST_INIT`.

**Review by Shadowtail117 — CHANGES_REQUESTED** [review]

File: code/game/world.dm (hunk 1 above)
Comment (Jack7D1) [reply code/game/world.dm:381]: good idea!

File: code/__DEFINES/misc.dm [hunk 2]
```diff
@@ -502,3 +502,19 @@ GLOBAL_LIST_INIT(pda_styles, sortList(list(MONO, VT, ORBITRON, SHARE)))
...
+//All new entries need to be added to the populate_bad_words proc in game/world.dm
+#define BAD_WORD_1_VAL  	"test"
+#define BAD_WORD_1_HASH 	"0391a1e58f324b3a0c79d32dd09436bd45bfc773"
+#define BAD_WORD_1_TARGET	"meatball"
+
+//Object-Position ownership map
+/*
+POSITION:		OBJECT PATH:
+1				/obj/item/reagent_containers/food/snacks/meatball
+End bad words*/
```
Comment (Shadowtail117) [inline code/__DEFINES/misc.dm:520]: So will we have to define `BAD_WORD_2_VAL BAD_WORD_2_HASH BAD_WORD_2_TARGET` `BAD_WORD_3_VAL _BAD_WORD3_HASH BARD_WORD_3_TARGET` here whenever we want to add new elements to the list?

Seems better to either make it a config thing or use global lists that are a bit more mutable than new defines for each entry.

Comment (Jack7D1) [reply code/__DEFINES/misc.dm:520]: Do you think three separate global lists corresponding to the same thing would be better? To me preprocessors seems a lit less heavy than adding new vars but if it'll be more convenient/clean then ye

Comment (Shadowtail117) [reply code/__DEFINES/misc.dm]: Bear in mind that these entries are defined *everywhere*. And I don't imagine there would be literally thousands of entries where performance would take a measurable hit.

Global lists are cleaner and generally easier to use than having to deal with many, many defines which you'd have to input manually into two places.

Comment (Jack7D1) [reply code/__DEFINES/misc.dm]: Implemented a subsystem specific 2d array!

**Comment by tiramisuapimancer** [comment]
this is a lot of work to be able to make meatballs called a slur again

**Comment by Maldaris** [comment]
> this is a lot of work to be able to make meatballs called a slur again

See this snippet from my comment above.

> That being said I do think this has some fun potential for encrypting strings for exploration content. This could be used to mildly obfuscate puzzles and content that would otherwise be readable/complete-able in singleplayer. Very interested to see where this goes.

**Comment by tiramisuapimancer** [comment]
yeah i read it i just feel off about the initial intent

**Comment by Jack7D1** [comment]
the power of autism is with me

**Comment by Jack7D1** [comment]
no slurs on this codebase at all

**Comment by AsciiSquid** [comment]
The initial intent is to test this tech on something simple and unobtrusive. Renaming a meatball to something we know can't be in the code anywhere is a start.

Shame people made the word for British meatballs naughty, however it's reintroduction shouldn't be something to take too seriously. It's all in good fun.

**Comment by dee-dubya** [comment]
the only thing that worries me about this rabbit hole is the potential of possibly obfuscating code from the maintainers themselves, sneaking in changes or features that we wouldn't want

**Comment by Sparrowbird611** [comment]
time to sneak ERP verbs into a PR :)

**Comment by AsciiSquid** [comment]
sadly this seems to only do strings, at best we can use this to obscure switch statements

and no ERP is still enforced. I **WILL** ban the code if I have to.

**Comment by Jack7D1** [comment]
This serves the use of obscuring certain recipes and other hints that may appear in game. Along with any general phrases that should not be able to be ctrl+f'able in our repo.

**Comment by Jack7D1** [comment]
warranty void if used for erp

**Review by MarkSuckerberg — APPROVED** [review]
Looks good :+1:

**Pull request merged**